### PR TITLE
Implemented shopping cart and items for that

### PR DESCRIPTION
**Pull request by Trayvor**

### CODE REVIEW

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public ShoppingCartDto getUsersShoppingCart() {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto getUsersShoppingCart(Authentication authentication) {
```

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        return shoppingCartService.getShoppingCart();
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public ShoppingCartDto addCartItemToShoppingCart(
+            @RequestBody CreateCartItemRequestDto createCartItemRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
            Authentication authentication,@RequestBody @Valid CreateCartItemRequestDto createCartItemRequestDto) {
```

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        return shoppingCartService.addItemToShoppingCart(createCartItemRequestDto);
+    }
+
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public UpdateCartItemResponseDto updateCartItemQuantity(@PathVariable Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemQuantity(Authentication authentication, @PathVariable Long cartItemId,
```

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                                                            @RequestBody UpdateCartItemRequestDto
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
                                                            @RequestBody @Valid UpdateCartItemRequestDto
```

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                                                          updateCartItemRequestDto) {
+        return cartItemService.updateCartItem(cartItemId, updateCartItemRequestDto);
+    }
+
+    @DeleteMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public void deleteCartItem(@PathVariable Long cartItemId) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public void deleteCartItem(Authentication authentication, @PathVariable Long cartItemId) {
```

File: src/main/java/org/example/bookstore/dto/cart/item/CreateCartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package org.example.bookstore.dto.cart.item;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/dto/cart/item/CreateCartItemRequestDto.java]: int can't be null
Add Positive if it's needed

File: src/main/java/org/example/bookstore/dto/cart/item/UpdateCartItemRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package org.example.bookstore.dto.cart.item;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/dto/cart/item/UpdateCartItemRequestDto.java]: same

File: src/main/java/org/example/bookstore/mapper/ShoppingCartMapper.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+@Mapper(config = MapperConfig.class)
+public interface ShoppingCartMapper {
+    ShoppingCart toModel(ShoppingCartDto shoppingCartDto);
+
+    default ShoppingCartDto toDto(ShoppingCart shoppingCart) {
+        ShoppingCartDto shoppingCartDto = new ShoppingCartDto();
+        shoppingCartDto.setId(shoppingCart.getId());
+        shoppingCartDto.setUserId(shoppingCart.getUser().getId());
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/mapper/ShoppingCartMapper.java]: create CartItem mapper and use it here

File: src/main/java/org/example/bookstore/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+@Getter
+@Setter
+@Entity
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @JoinColumn(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/org/example/bookstore/model/CartItem.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    private ShoppingCart shoppingCart;
+    @ManyToOne
+    @JoinColumn(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/org/example/bookstore/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/org/example/bookstore/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/model/ShoppingCart.java]: It's better to use

```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

File: src/main/java/org/example/bookstore/service/CartItemService.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package org.example.bookstore.service;
+
+import org.example.bookstore.dto.cart.item.UpdateCartItemRequestDto;
+import org.example.bookstore.dto.cart.item.UpdateCartItemResponseDto;
+
+public interface CartItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/service/CartItemService.java]: remove redundant service

File: src/main/java/org/example/bookstore/service/impl/CartItemServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import org.example.bookstore.exception.EntityNotFoundException;
+import org.example.bookstore.mapper.CartItemMapper;
+import org.example.bookstore.model.CartItem;
+import org.example.bookstore.repository.cart.item.CartItemRepository;
+import org.example.bookstore.service.CartItemService;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/service/impl/CartItemServiceImpl.java]: all these logic should be in shop cart service

File: src/main/java/org/example/bookstore/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -28,6 +34,22 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
+    @Override
+    public User getAuthenticatedUser() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        return userRepository.findByEmail(authentication.getName()).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find authenticated user with email "
+                        + authentication.getName())
+        );
+    }
+
+    private void createShoppingCartForUser(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/service/impl/UserServiceImpl.java]: it should be in shop cart service

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java (hunk 5 above)
Comment (Trayvor) [reply src/main/java/org/example/bookstore/controller/ShoppingCartController.java:50]: Why do we need Authentication object in delete endpoint?

File: src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 16]
```diff
@@ -58,4 +58,15 @@ public ResponseEntity<Object> handleRegistrationException(RegistrationException
 
         return new ResponseEntity<>(body, HttpStatus.CONFLICT);
     }
+
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex,
+                                                              WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put(TIMESTAMP_BODY, LocalDateTime.now());
+        body.put(STATUS_BODY, HttpStatus.CONFLICT);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java]: wrong status

File: src/main/java/org/example/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    public void deleteCartItem(Long itemId, Long userId) {
+        if (!cartItemRepository.existsByIdAndShoppingCartUserId(itemId, userId)) {
+            throw new EntityNotFoundException("Can`t find cart item with id " + itemId
+                    + ", because it does not exists or not belongs to user with id "
+                    + userId);
+        }
+        cartItemRepository.deleteById(itemId);
+    }
+
+    public void createShoppingCartForUser(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/service/impl/ShoppingCartServiceImpl.java]: Override

Comment (Elena-Bruyako) [inline src/main/java/org/example/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    public void createShoppingCart(User user) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Operation(summary = "Adding cart item",
+            description = "Adding cart item in user`s shopping cart")
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public ShoppingCartDto addCartItemToShoppingCart(
+            @RequestBody @Valid CreateCartItemRequestDto createCartItemRequestDto,
+            Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java]: User user = (User) authentication.getPrincipal(); extract repeated logic to a private method

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 19]
```diff
@@ -23,3 +23,11 @@ databaseChangeLog:
...
   - include:
       file: db/changelog/changes/12-insert-dependencies-into-books-categories-table.yaml
+  - include:
+      file: db/changelog/changes/13-create-shopping-carts-table.yaml
+  - include:
+      file: db/changelog/changes/14-create-cart-items-table.yaml
+  - include:
+      file: db/changelog/changes/15-insert-shopping-carts.yaml
+  - include:
+      file: db/changelog/changes/16-insert-cart-items.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/16-insert-cart-items.yaml

```

File: src/main/java/org/example/bookstore/model/ShoppingCart.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/model/ShoppingCart.java]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

File: src/main/java/org/example/bookstore/dto/cart/item/CreateCartItemRequestDto.java [hunk 21]
```diff
@@ -0,0 +1,13 @@
+package org.example.bookstore.dto.cart.item;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull
+    private Long bookId;
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/dto/cart/item/CreateCartItemRequestDto.java]:
```suggestion
    @Positive
    private long bookId;
```

**Review by okuzan — CHANGES_REQUESTED** [review]
Almost done

**Pull request merged**